File: kotlin/src/main/java/will/sudoku/solver/SolvingProgress.kt
package will.sudoku.solver

import kotlinx.serialization.Serializable

/**
 * Tracks the overall progress of solving a Sudoku puzzle step by step.
 *
 * This class accumulates all solving steps and provides:
 * - Step-by-step replay capability
 * - Progress statistics
 * - Serialization for API responses
 */
@Serializable
data class SolvingProgress(
    val originalPuzzle: String,
    val steps: MutableList<SolvingStep> = mutableListOf(),
    var currentBoardState: String,
    var isSolved: Boolean = false,
    var hasNoSolution: Boolean = false,
    var startTime: Long = System.currentTimeMillis(),
    var endTime: Long? = null
) {
    /**
     * Add a new step to the progress.
     */
    fun addStep(step: SolvingStep) {
        steps.add(step)
    }

    /**
     * Get the current step number (next step will be this + 1).
     */
    fun nextStepNumber(): Int = steps.size + 1

    /**
     * Mark the puzzle as solved.
     */
    fun markSolved(finalBoardState: String) {
        isSolved = true
        currentBoardState = finalBoardState
        endTime = System.currentTimeMillis()
        addStep(SolvingStep(
            stepNumber = nextStepNumber(),
            stepType = StepType.PUZZLE_SOLVED,
            affectedCells = emptyList(),
            values = emptySet(),
            explanation = "Puzzle solved successfully in ${steps.size} steps!"
        ))
    }

    /**
     * Mark the puzzle as having no solution.
     */
    fun markNoSolution(reason: String = "No valid solution exists") {
        hasNoSolution = true
        endTime = System.currentTimeMillis()
        addStep(SolvingStep(
            stepNumber = nextStepNumber(),
            stepType = StepType.NO_SOLUTION,
            affectedCells = emptyList(),
            values = emptySet(),
            explanation = reason
        ))
    }

    /**
     * Get total solving time in milliseconds.
     */
    fun solveTimeMs(): Long? {
        return if (endTime != null) endTime!! - startTime else null
    }

    /**
     * Get statistics about solving techniques used.
     */
    fun techniqueStats(): Map<StepType, Int> {
        return steps.groupingBy { it.stepType }.eachCount()
    }

    /**
     * Get steps filtered by technique.
     */
    fun stepsByType(type: StepType): List<SolvingStep> {
        return steps.filter { it.stepType == type }
    }

    /**
     * Get a summary of the solving process.
     */
    fun summary(): String {
        val timeStr = solveTimeMs()?.let { "${it}ms" } ?: "in progress"
        val stats = techniqueStats()
        val techniqueCounts = stats.entries
            .filter { it.key !in listOf(StepType.PUZZLE_SOLVED, StepType.NO_SOLUTION) }
            .joinToString(", ") { "${it.key.displayName}: ${it.value}" }

        return buildString {
            appendLine("Solving Progress")
            appendLine("─".repeat(40))
            appendLine("Status: ${when { isSolved -> "Solved ✓"; hasNoSolution -> "No Solution ✗"; else -> "In Progress..." }}")
            appendLine("Steps: ${steps.size}")
            appendLine("Time: $timeStr")
            if (techniqueCounts.isNotEmpty()) {
                appendLine("Techniques: $techniqueCounts")
            }
        }
    }

    companion object {
        /**
         * Create a new SolvingProgress from a puzzle string.
         */
        fun fromPuzzle(puzzle: String): SolvingProgress {
            return SolvingProgress(
                originalPuzzle = puzzle,
                currentBoardState = puzzle
            )
        }
    }
}

File: packages/solver/src/SolvingProgress.ts
import { StepType } from './StepType'
import type { SolvingStep } from './SolvingStep'

/**
 * Metrics collected for a single eliminator.
 *
 * Equivalent to the Kotlin EliminatorMetrics data class.
 */
export interface EliminatorMetrics {
  eliminations: number
  passes: number
  totalTimeNanos: number
}

/**
 * Metrics collected during Sudoku solving process.
 *
 * Equivalent to the Kotlin SolverMetrics data class.
 */
export interface SolverMetrics {
  totalSolveTimeNanos: number
  backtrackingCount: number
  maxRecursionDepth: number
  propagationPasses: number
  cellsProcessed: number
  eliminatorMetrics: Map<string, EliminatorMetrics>
}

/** Create default empty metrics. */
export function emptyMetrics(): SolverMetrics {
  return {
    totalSolveTimeNanos: 0,
    backtrackingCount: 0,
    maxRecursionDepth: 0,
    propagationPasses: 0,
    cellsProcessed: 0,
    eliminatorMetrics: new Map()
  }
}

/**
 * Tracks the overall progress of solving a Sudoku puzzle step by step.
 *
 * Equivalent to the Kotlin SolvingProgress data class.
 */
export class SolvingProgress {
  originalPuzzle: string
  steps: SolvingStep[] = []
  currentBoardState: string
  isSolved = false
  hasNoSolution = false
  startTime: number
  endTime?: number

  constructor(puzzle: string) {
    this.originalPuzzle = puzzle
    this.currentBoardState = puzzle
    this.startTime = Date.now()
  }

  addStep(step: SolvingStep): void {
    this.steps.push(step)
  }

  nextStepNumber(): number {
    return this.steps.length + 1
  }

  markSolved(finalBoardState: string): void {
    this.isSolved = true
    this.currentBoardState = finalBoardState
    this.endTime = Date.now()
    this.addStep({
      stepNumber: this.nextStepNumber(),
      stepType: StepType.PUZZLE_SOLVED,
      affectedCells: [],
      values: new Set(),
      explanation: `Puzzle solved successfully in ${this.steps.length} steps!`,
      timestamp: Date.now()
    })
  }

  markNoSolution(reason = 'No valid solution exists'): void {
    this.hasNoSolution = true
    this.endTime = Date.now()
    this.addStep({
      stepNumber: this.nextStepNumber(),
      stepType: StepType.NO_SOLUTION,
      affectedCells: [],
      values: new Set(),
      explanation: reason,
      timestamp: Date.now()
    })
  }

  solveTimeMs(): number | null {
    return this.endTime != null ? this.endTime - this.startTime : null
  }

  techniqueStats(): Map<StepType, number> {
    const stats = new Map<StepType, number>()
    for (const step of this.steps) {
      stats.set(step.stepType, (stats.get(step.stepType) ?? 0) + 1)
    }
    return stats
  }

  summary(): string {
    const timeStr = this.solveTimeMs() != null ? `${this.solveTimeMs()}ms` : 'in progress'
    const status = this.isSolved ? 'Solved ✓' : this.hasNoSolution ? 'No Solution ✗' : 'In Progress...'
    const stats = this.techniqueStats()
    const techniqueCounts = [...stats.entries()]
      .filter(([k]) => k !== StepType.PUZZLE_SOLVED && k !== StepType.NO_SOLUTION)
      .map(([k, v]) => `${k}: ${v}`)
      .join(', ')

    let s = `Solving Progress\n${'─'.repeat(40)}\n`
    s += `Status: ${status}\n`
    s += `Steps: ${this.steps.length}\n`
    s += `Time: ${timeStr}\n`
    if (techniqueCounts) s += `Techniques: ${techniqueCounts}\n`
    return s
  }
}
